doc/guide/guide-basics.rst

```diff
@@ -321,12 +323,43 @@ For the destruction operator above:
     False
 
     >>> q.data
+    Dia(shape=(4, 4), num_diag=1)
+
+
+The ``data`` attribute returns a Qutip diagonal matrix.
+``Qobj`` instances store their data in Qutip matrix format.
+In the core qutip module, the ``Dense``, ``CSR`` and ``Dia`` formats are available, but other module can add other formats.
+For example, the qutip-jax module add ``Jax`` and ``JaxDia`` formats.
+One can always access the underlying matrix as a numpy array using :meth:`.Qobj.full`.
+It is also possible to access the underlying data as is in a common format using :meth:`.Qobj.data_as`.
+
+.. doctest:: [basics]
+  :options: +NORMALIZE_WHITESPACE
+
+    >>> q.data_as("dia_matrix")
     <4x4 sparse matrix of type '<class 'numpy.complex128'>'
-	   with 3 stored elements in Compressed Sparse Row format>
+        with 3 stored elements (1 diagonals) in DIAgonal format>
 
 
+Conversion between storage type is done using the :meth:`.Qobj.to` method.
 
-The data attribute returns a message stating that the data is a sparse matrix. All ``Qobj`` instances store their data as a sparse matrix to save memory. To access the underlying dense matrix one needs to use the :func:`qutip.Qobj.full` function as described below.
+.. doctest:: [basics]
+  :options: +NORMALIZE_WHITESPACE
+
+    >>> q.to("CSR").data
+    CSR(shape=(4, 4), nnz=3)
+
+    >>> q.to("CSR").data_as("CSR_matrix")
+    <4x4 sparse matrix of type '<class 'numpy.complex128'>'
+        with 3 stored elements in Compressed Sparse Row format>
+
+
+Note that :meth:`.Qobj.data_as` does not do the conversion.
+
+QuTiP will do conversion when needed to keep everything working in any format.
+However these conversions could slow down the computations and it is recomented to keep to one family of format.
+For example, core qutip ``Dense`` and ``CSR`` work well together and binary operation between these format is efficient.
+However binary operations between ``Dense`` and ``Jax`` should be avoided since it is not clear whether the operation will be executed by Jax, (possibly on GPU) or numpy.
```

Comment:
```suggestion
However these conversions could slow down computation and it is recommended to keep to one format family where possible.
For example, core QuTiP  ``Dense`` and ``CSR`` work well together and binary operations between these formats is efficient.
However binary operations between ``Dense`` and ``Jax`` should be avoided since it is not always clear whether the operation will be executed by Jax (possibly on a GPU if present) or numpy.
```